packages/world/test.ts

```diff
@@ -35,55 +36,170 @@ describe("WorldReader", () => {
+            expect(region.Level.Entities).toBeInstanceOf(Array);
+            expect(region.Level.Sections).toBeInstanceOf(Array);
+        })
+        test("should be able to get the correct section", async () => {
+            let reader = await WorldReader.create(`${root}/saves/1.12.2`);
+            let region = await reader.getRegionData(0, 0);
+            let section = region.Level.Sections[0] as LegacyRegionSectionDataFrame;
+            // the blocks is the string format of nbt data in the `Block` section
+            let blocks = readFileSync(join(root, "saves/1.12.2/block.txt")).toString().split(" ").map((s) => s.trim()).filter((s) => s.length !== 0).map((s) => Number.parseInt(s));
+            // the data is the string format of nbt data in `Data` section
+            let data = readFileSync(join(root, "saves/1.12.2/data.txt")).toString().split(" ").map((s) => s.trim()).filter((s) => s.length !== 0).map((s) => Number.parseInt(s));
+
+            // If they equal, then we correctly found the section
+            expect(section.Blocks).toEqual(blocks);
+            expect(section.Data).toEqual(data);
+        });
+    });
 });
 
+
 describe("RegionReader", () => {
-    test("#getBlockStateAt", async () => {
-        const reader = await WorldReader.create(`${root}/saves/1.12`);
-        const region = await reader.getRegionData(0, 0);
-        // const regionReader = new RegionReader(region);
-        // const block = regionReader.getBlockStateAt(0, 0, 0);
-        // expect(block).toEqual({ Name: "minecraft:bedrock" });
-        const chunkY = 0;
-        // for (let chunkY = 0; chunkY < region.Level.Sections.length; chunkY++) {
-        const section = RegionReader.getSection(region, chunkY);
+    /**
+     * Convert Minecraft BlockState toString() result to object format.
+     */
+    function nameToBlockState(name: string) {
+        let patt = /Block\{(.+)\}(\[(.+)\])?/g;
+        let mat = patt.exec(name)!;
+        let result: any = {
+            Name: mat[1]!,
+        };
+        if (mat[2]!) {
+            result.Properties = {}
+            let s = mat[2].substring(1, mat[2]!.length - 1)!
+            s.split(",").map((s) => s.split("=")).forEach(([key, value]) => {
+                result.Properties[key] = value
+            });
+        }
+        return result;
+    }
 
-        // console.log(Reflect.get(section, "Palette"))
-        let arr: any = new Set();
-        RegionReader.readBlockState(section, (x, y, z, id) => {
-            if (y === 4) {
-                arr.add(id);
+    function getMockData(version: string) {
+        let expected: Record<number, { name: string; id: number }> = {};
+        readFileSync(`${root}/saves/${version}/mock.txt`).toString().split("\n").map((l) => l.split(" ")).forEach(([id, x, y, z, name]) => {
+            if (typeof x === "undefined") { return; }
+            let i = (Number.parseInt(y) << 8) | (Number.parseInt(z) << 4) | (Number.parseInt(x));
```

Comment:
Is this collapsing xyz coordinates into a 1D-array ordered YZX?

Reply:
Yes, Minecraft use this index in chunk

Reply:
Holy shit that's elegant. Although perhaps it should be a separate function to enhance readability? Could be probably be useful in other parts of the code as well.

Reply:
Actually, that function existed, and I just use it here.

Reply:
Hold on, did you just add that or did I completely overlook that?
If it's the second: my bad :)

Reply:
I added this function today and I just use it here in like 5 minutes ago